File: source/core/Controllers/csController.cs
﻿
using System;
using System.Reflection;
using System.Xml;
using System.Diagnostics;
using System.Text.RegularExpressions;
using System.Collections.Generic;
using Contensive.Core;
using Contensive.Core.Models.DbModels;
using Contensive.Core.Controllers;
using static Contensive.Core.Controllers.genericController;
using static Contensive.Core.constants;
using Contensive.BaseClasses;
//

namespace Contensive.Core {
    //
    //==========================================================================================
    /// <summary>
    /// A recordset-like construct
    /// </summary>
    public class csController : IDisposable {
        //
        private Contensive.Core.Controllers.coreController core;
        private int csPtr;
        private int openingMemberID;
        protected bool disposed = false;
        //
        // Constructor - Initialize the Main and Csv objects
        //
        public csController(coreController core) {
            this.core = core;
            openingMemberID = core.doc.sessionContext.user.id;
        }
        //
        //====================================================================================================
        /// <summary>
        /// dispose
        /// </summary>
        /// <param name="disposing"></param>
        protected virtual void Dispose(bool disposing) {
            if (!this.disposed) {
                if (disposing) {
                    //
                    // -- call .dispose for managed objects
                    if (csPtr > -1) {
                        core.db.csClose(ref csPtr);
                    }
                }
                //
                // -- Add code here to release the unmanaged resource.
            }
            this.disposed = true;
        }
        //
        //====================================================================================================
        /// <summary>
        /// insert
        /// </summary>
        /// <param name="ContentName"></param>
        /// <returns></returns>
        public bool insert(string ContentName) {
            bool success = false;
            try {
                if (csPtr != -1) {
                    core.db.csClose(ref csPtr);
                }
                csPtr = core.db.csInsertRecord(ContentName, openingMemberID);
                success = core.db.csOk(csPtr);
            } catch (Exception ex) {
                core.handleException(ex);
                throw;
            }
            return success;
        }
        //
        //====================================================================================================
        public bool openRecord(string ContentName, int recordId, string SelectFieldList = "", bool ActiveOnly = true) {
            bool success = false;
            try {
                if (csPtr != -1) {
                    core.db.csClose(ref csPtr);
                }
                csPtr = core.db.csOpen(ContentName, "id=" + recordId, "", ActiveOnly, 0, false, false, SelectFieldList, 1, 1);
                success = core.db.csOk(csPtr);
            } catch (Exception ex) {
                core.handleException(ex);
                throw;
            }
            return success;
        }
        //
        //====================================================================================================
        public bool open(string ContentName, string SQLCriteria = "", string SortFieldList = "", bool ActiveOnly = true, string SelectFieldList = "", int pageSize = 0, int PageNumber = 1) {
            bool success = false;
            try {
                if (csPtr != -1) {
                    core.db.csClose(ref csPtr);
                }
                csPtr = core.db.csOpen(ContentName, SQLCriteria, SortFieldList, ActiveOnly,0,false,false,SelectFieldList, pageSize, PageNumber);
                success = core.db.csOk(csPtr);
            } catch (Exception ex) {
                core.handleException(ex);
                throw;
            }
            return success;
        }
        //
        //====================================================================================================
        public bool openGroupUsers(List<string> GroupList, string SQLCriteria = "", string SortFieldList = "", bool ActiveOnly = true, int PageSize = 10, int PageNumber = 1) {
            bool success = false;
            try {
                if (csPtr != -1) {
                    core.db.csClose(ref csPtr);
                }
                csPtr = core.db.csOpenGroupUsers(GroupList, SQLCriteria, SortFieldList, ActiveOnly, PageSize, PageNumber);
                success = ok();
            } catch (Exception ex) {
                core.handleException(ex);
                throw;
            }
            return success;
        }
        //
        //====================================================================================================
        public bool openGroupUsers(string GroupName, string SQLCriteria = "", string SortFieldList = "", bool ActiveOnly = true, int PageSize = 10, int PageNumber = 1) {
            bool success = false;
            try {
                List<string> groupList = new List<string>();
                groupList.Add(GroupName);
                if (csPtr != -1) {
                    core.db.csClose(ref csPtr);
                }
                csPtr = core.db.csOpenGroupUsers(groupList, SQLCriteria, SortFieldList, ActiveOnly, PageSize, PageNumber);
                success = ok();
            } catch (Exception ex) {
                core.handleException(ex);
                throw;
            }
            return success;
        }
        //
        //====================================================================================================
        public bool openGroupListUsers(string GroupCommaList, string SQLCriteria = "", string SortFieldList = "", bool ActiveOnly = true, int PageSize = 10, int PageNumber = 1) {
            bool result = false;
            try {
                List<string> groupList = new List<string>();
                groupList.AddRange(GroupCommaList.Split(','));
                result = openGroupUsers(groupList, SQLCriteria, SortFieldList, ActiveOnly, PageSize, PageNumber);
            } catch (Exception ex) {
                core.handleException(ex);
                throw;
            }
            return result;
        }
        //
        //====================================================================================================
        public bool openSQL(string sql) {
            bool success = false;
            try {
                if (csPtr != -1) {
                    core.db.csClose(ref csPtr);
                }
                csPtr = core.db.csOpenSql_rev("default", sql);
                success = core.db.csOk(csPtr);
            } catch (Exception ex) {
                core.handleException(ex);
                throw;
            }
            return success;
        }
        //
        //====================================================================================================
        public bool openSQL(string sql, string DataSourcename, int PageSize = 10, int PageNumber = 1) {
            bool success = false;
            try {
                if (csPtr != -1) {
                    core.db.csClose(ref csPtr);
                }
                csPtr = core.db.csOpenSql(sql, DataSourcename, PageSize, PageNumber);
                success = core.db.csOk(csPtr);
            } catch (Exception ex) {
                core.handleException(ex);
                throw;
            }
            return success;
        }
        //
        //====================================================================================================
        public void Close() {
            try {
                if (csPtr != -1) {
                    core.db.csClose(ref csPtr);
                    csPtr = -1;
                }
            } catch (Exception ex) {
                core.handleException(ex);
                throw;
            }
        }
        //
        //====================================================================================================
        public string getFormInput(string ContentName, string FieldName, int HeightLines = 1, int WidthRows = 40, string HtmlId = "") {
            return core.html.inputCs(csPtr, ContentName, FieldName, HeightLines, WidthRows, HtmlId);
        }
        //
        //====================================================================================================
        public void delete() {
            core.db.csDeleteRecord(csPtr);
        }
        //
        //====================================================================================================
        public bool fieldOK(string FieldName) {
            return core.db.cs_isFieldSupported(csPtr, FieldName);
        }
        //
        //====================================================================================================
        public void goFirst() {
            core.db.cs_goFirst(csPtr, false);
        }
        //
        //====================================================================================================
        public string getAddLink(string PresetNameValueList = "", bool AllowPaste = false) {
            string result = "";
            try {
                result = core.html.cs_getRecordAddLink(csPtr, PresetNameValueList, AllowPaste);
                if (result == null) {
                    result = "";
                }
            } catch (Exception ex) {
                core.handleException(ex);
                throw;
            }
            return result;
        }
        //
        //====================================================================================================
        public bool getBoolean(string FieldName) {
            return core.db.csGetBoolean(csPtr, FieldName);
        }
        //
        //====================================================================================================
        public DateTime getDate(string FieldName) {
            return core.db.csGetDate(csPtr, FieldName);
        }
        //
        //====================================================================================================
        public string getEditLink(bool AllowCut = false) {
            string result = "";
            try {
                result = core.db.csGetRecordEditLink(csPtr, AllowCut);
                if (result == null) {
                    result = "";
                }
            } catch (Exception ex) {
                core.handleException(ex);
                throw;
            }
            return result;
        }
        //
        //====================================================================================================
        /// <summary>
        /// if the field is backed by a filename, use this method to read the filename
        /// </summary>
        /// <param name="FieldName"></param>
        /// <param name="OriginalFilename"></param>
        /// <param name="ContentName"></param>
        /// <param name="fieldTypeId"></param>
        /// <returns></returns>
        public string getFieldFilename(string FieldName, string OriginalFilename = "", string ContentName = "", int fieldTypeId = 0) {
            string result = "";
            try {
                result = core.db.csGetFieldFilename(csPtr, FieldName, OriginalFilename, ContentName, fieldTypeId);
                if (result == null) {
                    result = "";
                }
            } catch (Exception ex) {
                core.handleException(ex);
                throw;
            }
            return result;
        }
        //
        //====================================================================================================
        public int getInteger(string FieldName) {
            return core.db.csGetInteger(csPtr, FieldName);
        }
        //
        //====================================================================================================
        public double getNumber(string FieldName) {
            return core.db.csGetNumber(csPtr, FieldName);
        }
        //
        //====================================================================================================
        public int getRowCount() {
            return core.db.csGetRowCount(csPtr);
        }
        //
        //====================================================================================================
        public string getSql() {
            string result = "";
            try {
                result = core.db.csGetSource(csPtr);
                if (result == null) {
                    result = "";
                }
            } catch (Exception ex) {
                core.handleException(ex);
                throw;
            }
            return result;
        }
        //
        //====================================================================================================
        /// <summary>
        /// returns the text value stored in the field. For Lookup fields, this method returns the name of the foreign key record.
        /// For textFile fields, this method returns the filename.
        /// </summary>
        /// <param name="FieldName"></param>
        /// <returns></returns>
        public string getText(string FieldName) {
            string result = "";
            try {
                result = core.db.csGet(csPtr, FieldName);
                if (result == null) {
                    result = "";
                }
            } catch (Exception ex) {
                core.handleException(ex);
                throw;
            }
            return result;
        }
        //
        //====================================================================================================
        // Needs to be implemented (refactor to check the field type. if not fieldTypeHtml, encodeHtml)
        public string getHtml(string FieldName) {
            return getText(FieldName);
        }
        //
        //====================================================================================================
        // <summary>
        // returns the text stored in a textfile type field instead of the filename.
        // </summary>
        // <param name="FieldName"></param>
        // <returns></returns>
        //Public Function getTextFile(ByVal FieldName As String) As String
        //    Dim result As String = ""
        //    Try
        //        result = core.db.cs_getTextFile(csPtr, FieldName)
        //        If result Is Nothing Then
        //            result = ""
        //        End If
        //    Catch ex As Exception
        //        Call core.handleException(ex); : Throw
        //    End Try
        //    Return result
        //End Function
        //
        //====================================================================================================
        public void goNext() {
            core.db.csGoNext(csPtr);
        }
        //
        //====================================================================================================
        public bool nextOK() {
            bool result = false;
            try {
                core.db.csGoNext(csPtr);
                result = core.db.csOk(csPtr);
            } catch (Exception ex) {
                core.handleException(ex);
                throw;
            }
            return result;
        }
        //
        //====================================================================================================
        public bool ok() {
            return core.db.csOk(csPtr);
        }
        //
        //====================================================================================================
        public void save() {
            core.db.csSave2(csPtr);
        }
        //
        //====================================================================================================
        /// <summary>
        /// set the value for the field.
        /// </summary>
        /// <param name="FieldName"></param>
        /// <param name="FieldValue"></param>
        public void setField(string FieldName, DateTime FieldValue) {
            core.db.csSet(csPtr, FieldName, FieldValue);
        }
        //
        //====================================================================================================
        /// <summary>
        /// set the value for the field.
        /// </summary>
        /// <param name="FieldName"></param>
        /// <param name="FieldValue"></param>
        public void setField(string FieldName, bool FieldValue) {
            core.db.csSet(csPtr, FieldName, FieldValue);
        }
        //
        //====================================================================================================
        /// <summary>
        /// set the value for the field.
        /// </summary>
        /// <param name="FieldName"></param>
        /// <param name="FieldValue"></param>
        public void setField(string FieldName, string FieldValue) {
            core.db.csSet(csPtr, FieldName, FieldValue);
        }
        //
        //====================================================================================================
        /// <summary>
        /// set the value for the field.
        /// </summary>
        public void setField(string FieldName, double FieldValue) {
            core.db.csSet(csPtr, FieldName, FieldValue);
        }
        //
        //====================================================================================================
        /// <summary>
        /// set the value for the field. If the field is backed by a file, the value will be saved to the file
        /// </summary>
        /// <param name="FieldName"></param>
        /// <param name="FieldValue"></param>
        public void setField(string FieldName, int FieldValue) {
            core.db.csSet(csPtr, FieldName, FieldValue);
        }
        //
        //====================================================================================================
        /// <summary>
        /// if the field uses an underlying filename, use this method to set that filename. The content for the field will switch to that contained by the new file
        /// </summary>
        /// <param name="fieldName"></param>
        /// <param name="filename"></param>
        public void setFieldFilename( string fieldName, string filename ) {
            core.db.csSetFieldFilename(csPtr, fieldName, filename);
        }
        //
        //========================================================================
        //   main_cs_get Field, translate all fields to their best text equivalent, and encode for display
        //========================================================================
        //
        public static string getTextEncoded(coreController core, int CSPointer, string FieldName) {
            string ContentName = "";
            int RecordID = 0;
            if (core.db.cs_isFieldSupported(CSPointer, "id") & core.db.cs_isFieldSupported(CSPointer, "contentcontrolId")) {
                RecordID = core.db.csGetInteger(CSPointer, "id");
                ContentName = Models.Complex.cdefModel.getContentNameByID(core, core.db.csGetInteger(CSPointer, "contentcontrolId"));
            }
            string source = core.db.csGet(CSPointer, FieldName);
            return activeContentController.renderHtmlForWeb(core, source, ContentName, RecordID, core.doc.sessionContext.user.id, "", 0, CPUtilsBaseClass.addonContext.ContextPage);
        }
        //
        //========================================================================
        //
        public string getValue(string FieldName) {
            string result = "";
            try {
                result = core.db.cs_getValue(csPtr, FieldName);
                if (result == null) {
                    result = "";
                }
            } catch (Exception ex) {
                core.handleException(ex);
                throw;
            }
            return result;
        }

        //
        #region  IDisposable Support 
        // Do not change or add Overridable to these methods.
        // Put cleanup code in Dispose(ByVal disposing As Boolean).
        public void Dispose() {
            Dispose(true);
            GC.SuppressFinalize(this);
        }
        ~csController() {
            Dispose(false);
            //todo  NOTE: The base class Finalize method is automatically called from the destructor:
            //base.Finalize();
        }
        #endregion
    }

}

File: source/Processor/Controllers/7/csController.vb
﻿
Option Explicit On
Option Strict On

Imports Contensive.BaseClasses
Imports Contensive.Core.Controllers
Imports System.Runtime.InteropServices

Namespace Contensive.Core
    Public Class csController
        Implements IDisposable
        '
        Private cpCore As Contensive.Core.coreClass
        Private csPtr As Integer
        Private OpeningMemberID As Integer
        Protected disposed As Boolean = False
        '
        ' Constructor - Initialize the Main and Csv objects
        '
        Public Sub New(ByRef cpCore As coreClass)
            Me.cpCore = cpCore
        End Sub
        '
        '====================================================================================================
        ''' <summary>
        ''' dispose
        ''' </summary>
        ''' <param name="disposing"></param>
        Protected Overridable Overloads Sub Dispose(ByVal disposing As Boolean)
            If Not Me.disposed Then
                If disposing Then
                    '
                    ' -- call .dispose for managed objects
                    If csPtr > -1 Then
                        Call cpCore.db.csClose(csPtr)
                    End If
                End If
                '
                ' -- Add code here to release the unmanaged resource.
            End If
            Me.disposed = True
        End Sub
        '
        '====================================================================================================
        ''' <summary>
        ''' insert
        ''' </summary>
        ''' <param name="ContentName"></param>
        ''' <returns></returns>
        Public Function Insert(ByVal ContentName As String) As Boolean
            Dim success As Boolean = False
            Try
                If csPtr <> -1 Then
                    Call cpCore.db.csClose(csPtr)
                End If
                csPtr = cpCore.db.csInsertRecord(ContentName, OpeningMemberID)
                success = cpCore.db.csOk(csPtr)
            Catch ex As Exception
                Call cpCore.handleException(ex) : Throw
            End Try
            Return success
        End Function
        '
        '====================================================================================================
        Public Function OpenRecord(ByVal ContentName As String, ByVal recordId As Integer, Optional ByVal SelectFieldList As String = "", Optional ByVal ActiveOnly As Boolean = True) As Boolean
            Dim success As Boolean = False
            Try
                If csPtr <> -1 Then
                    Call cpCore.db.csClose(csPtr)
                End If
                csPtr = cpCore.db.csOpen(ContentName, "id=" & recordId, , ActiveOnly, , , , SelectFieldList, 1, 1)
                success = cpCore.db.csOk(csPtr)
            Catch ex As Exception
                Call cpCore.handleException(ex) : Throw
            End Try
            Return success
        End Function
        '
        '====================================================================================================
        Public Function open(ByVal ContentName As String, Optional ByVal SQLCriteria As String = "", Optional ByVal SortFieldList As String = "", Optional ByVal ActiveOnly As Boolean = True, Optional ByVal SelectFieldList As String = "", Optional ByVal pageSize As Integer = 0, Optional ByVal PageNumber As Integer = 1) As Boolean
            Dim success As Boolean = False
            Try
                If csPtr <> -1 Then
                    Call cpCore.db.csClose(csPtr)
                End If
                csPtr = cpCore.db.csOpen(ContentName, SQLCriteria, SortFieldList, ActiveOnly, , , , SelectFieldList, pageSize, PageNumber)
                success = cpCore.db.csOk(csPtr)
            Catch ex As Exception
                Call cpCore.handleException(ex) : Throw
            End Try
            Return success
        End Function
        '
        '====================================================================================================
        Public Function openGroupUsers(ByVal GroupList As List(Of String), Optional ByVal SQLCriteria As String = "", Optional ByVal SortFieldList As String = "", Optional ByVal ActiveOnly As Boolean = True, Optional ByVal PageSize As Integer = 10, Optional ByVal PageNumber As Integer = 1) As Boolean
            Dim success As Boolean = False
            Try
                If csPtr <> -1 Then
                    Call cpCore.db.csClose(csPtr)
                End If
                csPtr = cpCore.db.csOpenGroupUsers(GroupList, SQLCriteria, SortFieldList, ActiveOnly, PageSize, PageNumber)
                success = ok()
            Catch ex As Exception
                Call cpCore.handleException(ex) : Throw
            End Try
            Return success
        End Function
        '
        '====================================================================================================
        Public Function openGroupUsers(ByVal GroupName As String, Optional ByVal SQLCriteria As String = "", Optional ByVal SortFieldList As String = "", Optional ByVal ActiveOnly As Boolean = True, Optional ByVal PageSize As Integer = 10, Optional ByVal PageNumber As Integer = 1) As Boolean
            Dim success As Boolean = False
            Try
                Dim groupList As New List(Of String)
                groupList.Add(GroupName)
                If csPtr <> -1 Then
                    Call cpCore.db.csClose(csPtr)
                End If
                csPtr = cpCore.db.csOpenGroupUsers(groupList, SQLCriteria, SortFieldList, ActiveOnly, PageSize, PageNumber)
                success = ok()
            Catch ex As Exception
                Call cpCore.handleException(ex) : Throw
            End Try
            Return success
        End Function
        '
        '====================================================================================================
        Public Function openGroupListUsers(ByVal GroupCommaList As String, Optional ByVal SQLCriteria As String = "", Optional ByVal SortFieldList As String = "", Optional ByVal ActiveOnly As Boolean = True, Optional ByVal PageSize As Integer = 10, Optional ByVal PageNumber As Integer = 1) As Boolean
            Dim result As Boolean = False
            Try
                Dim groupList As New List(Of String)
                groupList.AddRange(GroupCommaList.Split(","c))
                result = openGroupUsers(groupList, SQLCriteria, SortFieldList, ActiveOnly, PageSize, PageNumber)
            Catch ex As Exception
                cpCore.handleException(ex) : Throw
            End Try
            Return result
        End Function
        '
        '====================================================================================================
        Public Function openSQL(ByVal sql As String) As Boolean
            Dim success As Boolean = False
            Try
                If csPtr <> -1 Then
                    Call cpCore.db.csClose(csPtr)
                End If
                csPtr = cpCore.db.csOpenSql_rev("default", sql)
                success = cpCore.db.csOk(csPtr)
            Catch ex As Exception
                Call cpCore.handleException(ex) : Throw
            End Try
            Return success
        End Function
        '
        '====================================================================================================
        Public Function openSQL(ByVal sql As String, ByVal DataSourcename As String, Optional ByVal PageSize As Integer = 10, Optional ByVal PageNumber As Integer = 1) As Boolean
            Dim success As Boolean = False
            Try
                If csPtr <> -1 Then
                    Call cpCore.db.csClose(csPtr)
                End If
                csPtr = cpCore.db.csOpenSql(sql, DataSourcename, PageSize, PageNumber)
                success = cpCore.db.csOk(csPtr)
            Catch ex As Exception
                Call cpCore.handleException(ex) : Throw
            End Try
            Return success
        End Function
        '
        '====================================================================================================
        Public Sub Close()
            Try
                If csPtr <> -1 Then
                    Call cpCore.db.csClose(csPtr)
                    csPtr = -1
                End If
            Catch ex As Exception
                Call cpCore.handleException(ex) : Throw
            End Try
        End Sub
        '
        '====================================================================================================
        Public Function getFormInput(ByVal ContentName As String, ByVal FieldName As String, Optional ByVal HeightLines As Integer = 1, Optional ByVal WidthRows As Integer = 40, Optional ByVal HtmlId As String = "") As String
            Return cpCore.html.html_GetFormInputCS(csPtr, ContentName, FieldName, HeightLines, WidthRows, HtmlId)
        End Function
        '
        '====================================================================================================
        Public Sub delete()
            Call cpCore.db.csDeleteRecord(csPtr)
        End Sub
        '
        '====================================================================================================
        Public Function fieldOK(ByVal FieldName As String) As Boolean
            Return cpCore.db.cs_isFieldSupported(csPtr, FieldName)
        End Function
        '
        '====================================================================================================
        Public Sub goFirst()
            Call cpCore.db.cs_goFirst(csPtr, False)
        End Sub
        '
        '====================================================================================================
        Public Function getAddLink(Optional ByVal PresetNameValueList As String = "", Optional ByVal AllowPaste As Boolean = False) As String
            Dim result As String = ""
            Try
                result = cpCore.html.main_cs_getRecordAddLink(csPtr, PresetNameValueList, AllowPaste)
                If result Is Nothing Then
                    result = String.Empty
                End If
            Catch ex As Exception
                Call cpCore.handleException(ex) : Throw
            End Try
            Return result
        End Function
        '
        '====================================================================================================
        Public Function getBoolean(ByVal FieldName As String) As Boolean
            Return cpCore.db.csGetBoolean(csPtr, FieldName)
        End Function
        '
        '====================================================================================================
        Public Function getDate(ByVal FieldName As String) As Date
            Return cpCore.db.csGetDate(csPtr, FieldName)
        End Function
        '
        '====================================================================================================
        Public Function getEditLink(Optional ByVal AllowCut As Boolean = False) As String
            Dim result As String = String.Empty
            Try
                result = cpCore.db.csGetRecordEditLink(csPtr, AllowCut)
                If result Is Nothing Then
                    result = String.Empty
                End If
            Catch ex As Exception
                Call cpCore.handleException(ex) : Throw
            End Try
            Return result
        End Function
        '
        '====================================================================================================
        Public Function getFilename(ByVal FieldName As String, Optional ByVal OriginalFilename As String = "", Optional ByVal ContentName As String = "", Optional fieldTypeId As Integer = 0) As String
            Dim result As String = String.Empty
            Try
                result = cpCore.db.csGetFilename(csPtr, FieldName, OriginalFilename, ContentName, fieldTypeId)
                If result Is Nothing Then
                    result = String.Empty
                End If
            Catch ex As Exception
                Call cpCore.handleException(ex) : Throw
            End Try
            Return result
        End Function
        '
        '====================================================================================================
        Public Function getInteger(ByVal FieldName As String) As Integer
            Return cpCore.db.csGetInteger(csPtr, FieldName)
        End Function
        '
        '====================================================================================================
        Public Function getNumber(ByVal FieldName As String) As Double
            Return cpCore.db.csGetNumber(csPtr, FieldName)
        End Function
        '
        '====================================================================================================
        Public Function getRowCount() As Integer
            Return cpCore.db.csGetRowCount(csPtr)
        End Function
        '
        '====================================================================================================
        Public Function getSql() As String
            Dim result As String = String.Empty
            Try
                result = cpCore.db.csGetSource(csPtr)
                If result Is Nothing Then
                    result = String.Empty
                End If
            Catch ex As Exception
                Call cpCore.handleException(ex) : Throw
            End Try
            Return result
        End Function
        '
        '====================================================================================================
        ''' <summary>
        ''' returns the text value stored in the field. For Lookup fields, this method returns the name of the foreign key record.
        ''' For textFile fields, this method returns the filename.
        ''' </summary>
        ''' <param name="FieldName"></param>
        ''' <returns></returns>
        Public Function getText(ByVal FieldName As String) As String
            Dim result As String = String.Empty
            Try
                result = cpCore.db.csGet(csPtr, FieldName)
                If result Is Nothing Then
                    result = String.Empty
                End If
            Catch ex As Exception
                Call cpCore.handleException(ex) : Throw
            End Try
            Return result
        End Function
        '
        '====================================================================================================
        ' Needs to be implemented (refactor to check the field type. if not fieldTypeHtml, encodeHtml)
        Public Function getHtml(ByVal FieldName As String) As String
            Return getText(FieldName)
        End Function
        ''
        ''====================================================================================================
        '''' <summary>
        '''' returns the text stored in a textfile type field instead of the filename.
        '''' </summary>
        '''' <param name="FieldName"></param>
        '''' <returns></returns>
        'Public Function getTextFile(ByVal FieldName As String) As String
        '    Dim result As String = String.Empty
        '    Try
        '        result = cpCore.db.cs_getTextFile(csPtr, FieldName)
        '        If result Is Nothing Then
        '            result = String.Empty
        '        End If
        '    Catch ex As Exception
        '        Call cpCore.handleException(ex) : Throw
        '    End Try
        '    Return result
        'End Function
        '
        '====================================================================================================
        Public Sub goNext()
            Call cpCore.db.csGoNext(csPtr)
        End Sub
        '
        '====================================================================================================
        Public Function nextOK() As Boolean
            Dim result As Boolean = False
            Try
                Call cpCore.db.csGoNext(csPtr)
                result = cpCore.db.csOk(csPtr)
            Catch ex As Exception
                Call cpCore.handleException(ex) : Throw
            End Try
            Return result
        End Function
        '
        '====================================================================================================
        Public Function ok() As Boolean
            Return cpCore.db.csOk(csPtr)
        End Function
        '
        '====================================================================================================
        Public Sub save()
            Call cpCore.db.csSave2(csPtr)
        End Sub
        '
        '====================================================================================================
        '
        Public Sub setField(ByVal FieldName As String, ByVal FieldValue As Date)
            Call cpCore.db.csSet(csPtr, FieldName, FieldValue)
        End Sub
        Public Sub setField(ByVal FieldName As String, ByVal FieldValue As Boolean)
            Call cpCore.db.csSet(csPtr, FieldName, FieldValue)
        End Sub
        Public Sub setField(ByVal FieldName As String, ByVal FieldValue As String)
            Call cpCore.db.csSet(csPtr, FieldName, FieldValue)
        End Sub
        Public Sub setField(ByVal FieldName As String, ByVal FieldValue As Double)
            Call cpCore.db.csSet(csPtr, FieldName, FieldValue)
        End Sub
        Public Sub setField(ByVal FieldName As String, ByVal FieldValue As Integer)
            Call cpCore.db.csSet(csPtr, FieldName, FieldValue)
        End Sub
        '
        '====================================================================================================
        Public Sub setFile(ByVal FieldName As String, ByVal Copy As String, ByVal ContentName As String)
            Call cpCore.db.csSetTextFile(csPtr, FieldName, Copy, ContentName)
        End Sub
        '
        '====================================================================================================
        Public Sub SetFormInput(ByVal FieldName As String, Optional ByVal RequestName As String = "")
            Call cs_setFormInput(cpCore, csPtr, FieldName, RequestName)
        End Sub
        '
        '
        '
        Public Shared Sub cs_setFormInput(cpcore As coreClass, ByVal CSPointer As Integer, ByVal FieldName As String, Optional ByVal RequestName As String = "")
            Dim LocalRequestName As String
            Dim Filename As String
            Dim Path As String
            '
            'If Not (true) Then Exit Sub
            '
            If Not cpcore.db.csOk(CSPointer) Then
                Throw New ApplicationException("ContentSetPointer is invalid, empty, or end-of-file")
            ElseIf Trim(FieldName) = "" Then
                Throw New ApplicationException("FieldName is invalid or blank")
            Else
                LocalRequestName = RequestName
                If LocalRequestName = "" Then
                    LocalRequestName = FieldName
                End If
                Select Case cpcore.db.cs_getFieldTypeId(CSPointer, FieldName)
                    Case FieldTypeIdBoolean
                        '
                        ' Boolean
                        '
                        Call cpcore.db.csSet(CSPointer, FieldName, cpcore.docProperties.getBoolean(LocalRequestName))
                    Case FieldTypeIdCurrency, FieldTypeIdFloat, FieldTypeIdInteger, FieldTypeIdLookup, FieldTypeIdManyToMany
                        '
                        ' Numbers
                        '
                        Call cpcore.db.csSet(CSPointer, FieldName, cpcore.docProperties.getNumber(LocalRequestName))
                    Case FieldTypeIdDate
                        '
                        ' Date
                        '
                        Call cpcore.db.csSet(CSPointer, FieldName, cpcore.docProperties.getDate(LocalRequestName))
                    Case FieldTypeIdFile, FieldTypeIdFileImage
                        '
                        '
                        '
                        Filename = cpcore.docProperties.getText(LocalRequestName)
                        If Filename <> "" Then
                            Path = cpcore.db.csGetFilename(CSPointer, FieldName, Filename,, cpcore.db.cs_getFieldTypeId(CSPointer, FieldName))
                            Call cpcore.db.csSet(CSPointer, FieldName, Path)
                            Path = genericController.vbReplace(Path, "\", "/")
                            Path = genericController.vbReplace(Path, "/" & Filename, "")
                            Call cpcore.appRootFiles.upload(LocalRequestName, Path, Filename)
                        End If
                    Case Else
                        '
                        ' text files
                        '
                        Call cpcore.db.csSet(CSPointer, FieldName, cpcore.docProperties.getText(LocalRequestName))
                End Select
            End If
        End Sub
        '
        '========================================================================
        '   main_cs_get Field, translate all fields to their best text equivalent, and encode for display
        '========================================================================
        '
        Public Shared Function getTextEncoded(cpcore As coreClass, ByVal CSPointer As Integer, ByVal FieldName As String) As String
            Dim ContentName As String = String.Empty
            Dim RecordID As Integer = 0
            If cpcore.db.cs_isFieldSupported(CSPointer, "id") And cpcore.db.cs_isFieldSupported(CSPointer, "contentcontrolId") Then
                RecordID = cpcore.db.csGetInteger(CSPointer, "id")
                ContentName = models.complex.cdefmodel.getContentNameByID(cpcore,cpcore.db.csGetInteger(CSPointer, "contentcontrolId"))
            End If
            Dim source As String = cpcore.db.csGet(CSPointer, FieldName)
            Return cpcore.html.convertActiveContentToHtmlForWebRender(source, ContentName, RecordID, cpCore.doc.authContext.user.id, "", 0, CPUtilsBaseClass.addonContext.ContextPage)
            'Return cpcore.html.convertActiveContent_internal(source, cpCore.doc.authContext.user.id, ContentName, RecordID, 0, False, False, True, True, False, True, "", "http://" & cpcore.webServer.requestDomain, False, 0, "", CPUtilsBaseClass.addonContext.ContextPage, cpCore.doc.authContext.isAuthenticated, Nothing, cpCore.doc.authContext.isEditingAnything())
        End Function
        '
        '========================================================================
        '
        Public Function getValue(ByVal FieldName As String) As String
            Dim result As String = String.Empty
            Try
                result = cpCore.db.cs_getValue(csPtr, FieldName)
                If result Is Nothing Then
                    result = String.Empty
                End If
            Catch ex As Exception
                Call cpCore.handleException(ex) : Throw
            End Try
            Return result
        End Function

        '
#Region " IDisposable Support "
        ' Do not change or add Overridable to these methods.
        ' Put cleanup code in Dispose(ByVal disposing As Boolean).
        Public Overloads Sub Dispose() Implements IDisposable.Dispose
            Dispose(True)
            GC.SuppressFinalize(Me)
        End Sub
        Protected Overrides Sub Finalize()
            Dispose(False)
            MyBase.Finalize()
        End Sub
#End Region
    End Class

End Namespace
